### PR TITLE
Post process html to include teardown in log

### LINKED ISSUE #131: Report fixture setup and teardown
Hi,
I would like to vizualize when fixtures are set up and teared down, the time they take, their output and if they failed.
I think they could be added in the Results section, filtered by default, around the tests.
It might add a lot of information, so it would look better along with #33 

Would you be open to such an enchancement? (It could be optional if needed.)
Do you see issues it might create? Would you do it in a different way?

### PR DESCRIPTION
**Pull request by csm10495**
I believe this would resolve #131 ... however it isn't really optimal, I didn't want to dig into the html logic, so instead just moved it to be post processed and tried to coerce the test result into what the existing code seemed to want.

### CODE REVIEW

**Comment by csm10495** [comment]
Wow. Those unit tests really helped me fix a lot of things i had broken :P. 

Honestly don't know if this is good enough to merge... it seems kind of sketchy... but per the tests seems to work. :)

**Comment by BeyondEvil** [comment]
Thanks for this! @csm10495 

I need some time to fully understand the problem _and_ the fix before I can review this with confidence.

Feel free to ping me if you haven't heard anything from me in a week or so.

**Comment by csm10495** [comment]
Sure thing @BeyondEvil . The big thing that I want to fix is to get the teardown event in the html log:

for example:


```
import pytest
pytest.fixture(scope='function')
def setupAndTeardown():
    print ("setup: i am already in the html")
    yield
    print ("teardown: i am missing from the html... but added by this change!")

def test_nothing():
    print ("test_case: of course i am already in the html")
```

In doing that, i try to merge hold back from processing the reports, until we get to the ```pytest_sessionfinish()``` then basically combine the setup/testcase/teardown reports to figure out the status, etc.

**Comment by csm10495** [comment]
@BeyondEvil reminder to check this out. Thanks.

**Comment by csm10495** [comment]
Reminder to check this out @BeyondEvil

**Comment by dasm** [comment]
@csm10495 @BeyondEvil I'd like to help with that functionality. Do you need anything extra?

Interesting fact. I'm able to see `captured stdout teardown` in HTML if teardown failed.

**Comment by Formartha** [comment]
Hi folks, any update about this one? we're waiting for it as well :).
a W/A can be provided in case you have one till it will be merged..

**Comment by dina809** [comment]
Hi @csm10495 - i have tried your code and it works good while collecting logs from teardown of a fixture and showing them in the html report,
but it is causing an issue if the fixture setup is failing (before the yield), have you checked that?

**Comment by csm10495** [comment]
Hey @dina809 , it's been a while now since I've looked deep at this. What error are you seeing? It works if I do this:

```
import pytest
@pytest.fixture(scope="function")
def setupAndTeardown():
    print ("this is setup")
    assert 0
    yield
    print ("this is teardown")
def test_setup_and_teardown(setupAndTeardown):
    print ("this is the test case")
```

HTML:
image

**Comment by avzero07** [comment]
Bump.

**Comment by 1Mark** [comment]
Any ETA for release date for this?

**Comment by BeyondEvil** [comment]
> Any release date for this?

I'm hoping @ssbarnea  or @gnikonorov can find some time to review this?

**Comment by gnikonorov** [comment]
I can take a look at this tonight, @BeyondEvil

File: pytest_html/plugin.py
Comment (gnikonorov) [inline pytest_html/plugin.py:278]: Can you add a comment explaining why `longreprtext` is Falsey at times and why we default to `full_text` in those instances?

File: pytest_html/plugin.py
Comment (gnikonorov) [inline pytest_html/plugin.py:631]: Can this ever be `None`? I ask since you add the following on line 278: `text = report.longreprtext or report.full_text`. Should we use `getattr` like we do below?

File: pytest_html/plugin.py
Comment (gnikonorov) [inline pytest_html/plugin.py:648]: Should lines 648->665 be indented in one tab so that it's in the for loop?

**Review by gnikonorov — CHANGES_REQUESTED** [review]
Overall this looks good, thank you @csm10495 and sorry for the very long delay here! I just had a few questions.

I'll be staying on top of this PR. If you haven't heard from me in a day or so, @ mention me so I see the notification

**Comment by csm10495** [comment]
@gnikonorov, I made some fixes and added a comment. Tests pass feel free to merge if its good.

Though I will say that I wrote this a long time ago for a previous task, so I don't quite have the right mindset to fully test it again. Hoping the new unit tests (that i wrote back then) are good enough.

File: pytest_html/plugin.py
Comment (gnikonorov) [reply pytest_html/plugin.py:658]: even though this is valid, can you either add a comment stating this is intentional or use a different variable name? I thought it was a typo at first glance

File: pytest_html/plugin.py
Comment (gnikonorov) [inline pytest_html/plugin.py:665]: can we simplify this to `test_report.wasxfail = wasxfail`?

Or can `test_report.wasxfail` still be `True` even if `wasxfail` is `False`?

File: pytest_html/plugin.py
Comment (gnikonorov) [inline pytest_html/plugin.py:672]: Shouldn't we call `self.append_other(test_report)` as we were doing in the `else` case for `pytest_runtest_logreport`?

**Review by gnikonorov — CHANGES_REQUESTED** [review]
Thanks for the prompt update, @csm10495! I had one more batch of comments.

Once those are addressed, I'd be happy to merge assuming @ssbarnea can take a look as well and OK this.

Also, can you add a changelog entry for this under `2.1.2 (unreleased)` in https://github.com/pytest-dev/pytest-html/blob/master/CHANGES.rst

File: pytest_html/plugin.py
Comment (csm10495) [reply pytest_html/plugin.py:672]: added comment for it

File: pytest_html/plugin.py
Comment (csm10495) [reply pytest_html/plugin.py:665]: It can't because other code relies on the attribute existing or not. So if we do the suggested, it would be false and exist.. and lead to more changes needed than desired to go into for this changeset.

Comment (gnikonorov) [reply pytest_html/plugin.py:665]: Thanks for clarifying @csm10495, I forgot about that

**Comment by csm10495** [comment]
@gnikonorov , changes made. Hoping we're all good :)

**Review by gnikonorov — DISMISSED** [review]
> @gnikonorov , changes made. Hoping we're all good :)

We are @csm10495, thank you for the prompt responses. I'll wait for @ssbarnea to take a look, but this LGTM to me

**Comment by ssbarnea** [comment]
Please rebase. I do support the change, is very useful to have teardowns with results.

**Comment by csm10495** [comment]
Thanks @ssbarnea and @gnikonorov. Squashed things down. Should be clean for merge now.

**Review by gnikonorov — APPROVED** [review]
LGTM, I'll leave the  merge to @ssbarnea so he can take a look at the final version too

**Review by ssbarnea — APPROVED** [review]

**Pull request merged**
